File: objectWithTypeScript/src/main/Chapter01/Bag.ts
import { Invitation } from './Invitation';
import { Ticket } from './Ticket';

export class Bag {
  private _amount: number;
  private _invitation: Invitation;
  private _ticket: Ticket;


  constructor(amount: number, invitation: Invitation = null) {
    this._amount = amount;
    this._invitation = invitation;
  }

  get amount(): number {
    return this._amount;
  }

  hasInvitation(): boolean {
    return this._invitation != null;
  }

  hasTicket(): boolean {
    return this._ticket != null;
  }

  setTicket(ticket: Ticket): void {
    this._ticket = ticket;
  }

  minusAmount(amount: number): void {
    this._amount -= amount;
  }

  plusAmount(amount: number): void {
    this._amount += amount;
  }
}

File: objectWithJava/src/main/com/chapter01/Bag.java
package com.chapter01;

public class Bag {
  private Long amount;
  private Invitation invitation;
  private Ticket ticket;
  
  public Bag(long amount) {
    this(null, amount);
  }
  
  public Bag(Invitation invitation, long amount) {
    this.invitation = invitation;
    this.amount = amount;
  }
  
  public Long getAmount() {
    return amount;
  }
  
  public boolean hasInvitation() {
    return invitation != null;
  }
  
  public boolean hasTicket() {
    return ticket != null;
  }
  
  public void setTicket(Ticket ticket) {
    this.ticket = ticket;
  }
  
  public void minusAmount(Long amount) {
    this.amount -= amount;
  }
  
  public void plusAmount(Long amount) {
    this.amount += amount;
  }
  
  public Long hold(Ticket ticket) {
    if (this.hasInvitation()) {
      this.setTicket(ticket);
      return 0L;
    } else {
      this.minusAmount(ticket.getFee());
      this.setTicket(ticket);
      return ticket.getFee();
    }
  }
}
